Sources/NIO/Thread.swift

```diff
@@ -117,6 +117,11 @@ final class Thread {
         return pthread_equal(pthread, pthread_self()) != 0
     }
 
+    func join() {
+        // We ignore the return value as the pthread may already be dead.
+        pthread_join(pthread, nil)
```

Comment:
Should we add a precondition here that confirms that we aren't joining on ourselves?

Reply:
If you are adding support for joining threads, shouldn't you also stop detaching in spawnAndRun?

Reply:
Let me make it an assert

Reply:
@toffaletti good Point....  @jw @Lukasa agree ?

Reply:
I'm certainly happy with this change & `assert`. Should be documented then (and also `preconditioned` in `Thread.deinit` that the thread has been joined

Reply:
actually, that has one problem (which is certainly API breaking): previously we didn't actually require you to shut down the event loops (as we had detached threads and we never joined). With that change we would require users to shut the event loops down properly in order for us to join the threads. And it's required to join non-detached threads or else Thread sanitizer will barf.

Reply:
you mean the check in deinit would be a breakage currect @weissi ? Inthinkwe can find everything except the deinit change

Reply:
Beside this we could also just revert the join but and use the while loop in the test as before. At the end it was mainly to make the test better

Reply:
@normanmaurer yes. If we decide to not detach the threads anymore, we have two options (both of which break):

1) we do put a `precondition` in `deinit` that a `Thread` has been `join`ed before being released. That'd break as currently you're not required to do that.
2) we do _not_ put a `precondition` in `deinit`: That has the same effect when run under TSan as TSan barks if you leaks a thread that is running (unless it's detached).

In other words, if we don't detach our threads we have a problem as we currently don't require users to shut down the event loops (which is a mistake).

Reply:
oh, and as @toffaletti points out, we can't `pthread_join` a thread that is detached. So I think we're stuck 😤

Reply:
@normanmaurer yes, I think we have to back off with the joining :|. We should however make this change in 2.0.0.

Reply:
Ok let me use the while loop again and add a comment that we should use join in the next major